src/main/kotlin/org/jetbrains/research/libsl/nodes/references/TypeReference.kt

```diff
@@ -3,55 +3,152 @@ package org.jetbrains.research.libsl.nodes.references
 import org.jetbrains.research.libsl.context.LslContextBase
 import org.jetbrains.research.libsl.type.*
 
-abstract class TypeReference(
-    open val name: String,
-    open val isPointer: Boolean = false,
-    open var typeBound: GenericTypeBound = GenericTypeBound.EMPTY,
-    open val genericReferences: MutableList<TypeReference> = mutableListOf(),
+interface TypeReference : LslReference<Type, TypeReference> {
     override val context: LslContextBase
-) : LslReference<Type, TypeReference> {
+}
+
+data class UnionExpressionTypeReference(
+    val left: TypeReference,
+    val right: TypeReference,
+    override val context: LslContextBase
+) : TypeReference {
+    override fun resolve(): Type? {
+        this.left.resolve()
+        this.right.resolve()
+        // We don't have type for UnionExpression
+        return null
+    }
+
+    override fun isReferenceMatchWithNode(node: Type): Boolean {
+        // we don't have Type for UnionExpressionTypeReference
+        return false
+    }
+
+    override fun isSameReference(other: TypeReference): Boolean {
+        // This is right realization ??
+        if (other is UnionExpressionTypeReference) {
+            this.left.isSameReference(other.left) && this.right.isSameReference(other.right)
+        }
+        return false
+    }
+
+    fun getName(): String {
+        return this.left.getName() + " | " + this.right.getName()
+    }
+
+}
+
+data class IntersectionExpressionTypeReference(
+    val left: TypeReference,
+    val right: TypeReference,
+    override val context: LslContextBase
+) : TypeReference {
+    override fun resolve(): Type? {
+        this.left.resolve()
+        this.right.resolve()
+        // We don't have type for IntersectionExpression
+        return null
+    }
+
+    override fun isReferenceMatchWithNode(node: Type): Boolean {
+        // we don't have Type for IntersectionExpressionTypeReference
+        return false
+    }
+
+    override fun isSameReference(other: TypeReference): Boolean {
+        // This is right realization ??
+        if (other is IntersectionExpressionTypeReference) {
+            this.left.isSameReference(other.left) && this.right.isSameReference(other.right)
+        }
+        return false
+    }
+
+    fun getName(): String {
+        return this.left.getName() + " & " + this.right.getName()
+    }
+}
+
+data class LiteralTypeReference(
+    val value: String,
```

Comment:
I suggest saving an ACTUAL (parsed) value here, not a string representation of it.
```suggestion
    val value: object,
```

Reply:
@dog-m maybe "any" not "object" ?

Reply:
Indeed `Any` (Kotlin), not an `object` (Java). Sorry.

Reply:
Fixed;